nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/PasswordBasedOauth2TokenProvider.java

```diff
@@ -0,0 +1,230 @@
+            try {
+                refreshAuthorizationDetails();
+            } catch (Exception e) {
+                getLogger().info("Couldn't refresh access token, probably refresh token has expired." +
+                    " Getting new token using credentials.");
+                acquireAuthorizationDetails();
+            }
+        }
+        accessToken = accessDetails.getAccessToken();
+
+        return accessToken;
+    }
+
+    private void acquireAuthorizationDetails() {
+        FormBody.Builder acquireTokenBuilder = new FormBody.Builder()
+            .add("grant_type", "password")
+            .add("username", username)
+            .add("password", password);
+
+        if (clientId != null) {
+            acquireTokenBuilder.add("client_id", clientId);
+            acquireTokenBuilder.add("client_secret", clientSecret);
+        }
+
+        RequestBody acquireTokenRequestBody = acquireTokenBuilder.build();
+
+        Request acquireTokenRequest = new Request.Builder()
+            .url(authorizationServerUrl)
+            .post(acquireTokenRequestBody)
+            .build();
+
+        this.accessDetails = getAccessDetails(acquireTokenRequest);
+    }
+
+    private void refreshAuthorizationDetails() {
+        FormBody.Builder refreshTokenBuilder = new FormBody.Builder()
+            .add("grant_type", "refresh_token")
+            .add("refresh_token", this.accessDetails.getRefreshToken());
+
+        if (clientId != null) {
+            refreshTokenBuilder.add("client_id", clientId);
+            refreshTokenBuilder.add("client_secret", clientSecret);
+        }
+
+        RequestBody refreshTokenRequestBody = refreshTokenBuilder.build();
+
+        Request refreshRequest = new Request.Builder()
+                .url(authorizationServerUrl)
+                .post(refreshTokenRequestBody)
+                .build();
+
+        this.accessDetails = getAccessDetails(refreshRequest);
+    }
+
+    private AccessDetails getAccessDetails(Request newRequest) {
+        try {
+            Response response = httpClient.newCall(newRequest).execute();
+            String responseBody = response.body().string();
+            if (response.code() != 200) {
+                getLogger().error(String.format("Bad response from the server during oauth2 request:\n%s", responseBody));
+                throw new ProcessException(String.format("Got HTTP %d during oauth2 request.", response.code()));
+            }
+
+            ObjectMapper mapper = new ObjectMapper()
+                .configure(com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+                .setPropertyNamingStrategy(com.fasterxml.jackson.databind.PropertyNamingStrategies.SNAKE_CASE);
+
+            AccessDetails accessDetails = mapper.readValue(responseBody, AccessDetails.class);
+
+            return accessDetails;
+        } catch (IOException e) {
+            throw new ProcessException(e);
```

Comment:
It seems like another more specific exception could be better here, or perhaps an `UncheckedIOException`. In either case, a message should be included for troubleshooting. In this case, the `IOException` could be coming from the HTTP call itself, or the JSON parsing. Recommend breaking out those operations into separate methods for clearer error handling.

Reply:
I'm not sure I see the value in making this more complex.
Whatever the issue was it will get reported properly and no information will be lost. I think we can rely on the default exception handling of the framework.

Reply:
For clarification, I am not recommending making this more complex, but just changing the wrapping exception. Since this is catching `IOException`, wrapping it in an `UncheckedIOException`, with a message, seems better.